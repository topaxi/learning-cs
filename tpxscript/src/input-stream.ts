export class InputStream {
  private pos = 0
  private line = 1
  private col = 0

  constructor(private input: string) {}

  next(): string {
    let ch = this.input.charAt(this.pos++)

    if (ch === '\n') {
      this.line++
      this.col = 0
    } else {
      this.col++
    }

    return ch
  }

  peek(): string {
    return this.input.charAt(this.pos)
  }

  eof(): boolean {
    return this.peek() === ''
  }

  croak(msg: string): never {
    throw new Error(`${msg} (${this.line}:${this.col})`)
  }
}
