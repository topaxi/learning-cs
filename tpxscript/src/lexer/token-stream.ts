import { InputStream } from '../input-stream'
import * as t from './tokens'
import w from 'w-array'

export class TokenStream {
  private current: t.Token | null = null
  private keywords = new Set(w`
    if
    then
    else
    fn
    true
    false
    and
    or
  `)

  constructor(readonly input: InputStream) {}

  isKeyword(ch: string): boolean {
    return this.keywords.has(ch)
  }

  isDigit(ch: string): boolean {
    return /\d/.test(ch)
  }

  isIdentifierStart(ch: string): boolean {
    return /[a-z_]/i.test(ch)
  }

  isIdentifier(ch: string): boolean {
    return this.isIdentifierStart(ch) || '?!-<>=0123456789'.includes(ch)
  }

  isOperatorCharacter(ch: string): boolean {
    return '+-*/%=&|<>!'.includes(ch)
  }

  isPunctuation(ch: string): boolean {
    return ',;(){}[]'.includes(ch)
  }

  isWhitespace(ch: string): boolean {
    return /\s/.test(ch)
  }

  readWhile(predicate: (this: this, ch: string) => boolean): string {
    let str = ''

    while (!this.input.eof() && predicate.call(this, this.input.peek())) {
      str = str + this.input.next()
    }

    return str
  }

  readNumber(): t.NumberToken {
    let hasDot = false
    let n = this.readWhile(ch => {
      if (ch === '.') {
        if (hasDot) return false
        hasDot = true
        return true
      }
      return this.isDigit(ch)
    })

    return { type: 'num', value: Number.parseFloat(n) }
  }

  readIdentifier(): t.IdentifierToken {
    let id = this.readWhile(this.isIdentifier)

    return { type: this.isKeyword(id) ? 'kw' : 'var', value: id }
  }

  readEscaped(end: string): string {
    let escaped = false
    let str = ''

    this.input.next()

    while (!this.input.eof()) {
      let ch = this.input.next()

      if (escaped) {
        str = str + ch
        escaped = false
      } else if (ch === '\\') {
        escaped = true
      } else if (ch === end) {
        break
      } else {
        str = str + ch
      }
    }

    return str
  }

  readString(): t.StringToken {
    return { type: 'str', value: this.readEscaped('"') }
  }

  skipComment(): void {
    this.readWhile(ch => ch !== '\n')
    this.input.next()
  }

  readNext(): t.Token | null {
    this.readWhile(this.isWhitespace)

    if (this.input.eof()) return null

    let ch = this.input.peek()

    if (ch === '#') {
      this.skipComment()
      return this.readNext()
    }

    if (ch === '"') return this.readString()
    if (this.isDigit(ch)) return this.readNumber()
    if (this.isIdentifierStart(ch)) return this.readIdentifier()
    if (this.isPunctuation(ch))
      return { type: 'punc', value: this.input.next() }
    if (this.isOperatorCharacter(ch))
      return { type: 'op', value: this.readWhile(this.isOperatorCharacter) }

    return this.input.croak(`Unable to handle character: ${ch}`)
  }

  peek(): t.Token | null {
    return this.current || (this.current = this.readNext())
  }

  next(): t.Token | null {
    let token = this.current
    this.current = null
    return token || this.readNext()
  }

  eof(): boolean {
    return this.peek() === null
  }

  croak(msg: string): never {
    return this.input.croak(msg)
  }
}
