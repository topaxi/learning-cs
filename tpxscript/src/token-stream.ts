import { InputStream } from './input-stream'

export type TokenType =
  | 'num'
  | 'kw'
  | 'punc'
  | 'op'
  | 'str'
  | 'bool'
  | 'var'
  | 'binary'
  | 'assign'
  | 'let'
  | 'fn'
  | 'if'
  | 'prog'
  | 'call'

export type Token<T = any> = { type: TokenType; value: T }

export class TokenStream {
  private current: Token | null = null
  private keywords = 'if then else fn true false'

  constructor(readonly input: InputStream) {}

  isKeyword(ch: string): boolean {
    return this.keywords.includes(ch)
  }

  isDigit(ch: string): boolean {
    return /\d/.test(ch)
  }

  isIdentifierStart(ch: string): boolean {
    return /[a-z_]/i.test(ch)
  }

  isIdentifier(ch: string): boolean {
    return this.isIdentifierStart(ch) || '?!-<>=0123456789'.includes(ch)
  }

  isOperatorCharacter(ch: string): boolean {
    return '+-*/%=&|<>!'.includes(ch)
  }

  isPunctuation(ch: string): boolean {
    return ',;(){}[]'.includes(ch)
  }

  isWhitespace(ch: string): boolean {
    return /\s/.test(ch)
  }

  readWhile(predicate: (this: this, ch: string) => boolean): string {
    let str = ''

    while (!this.input.eof() && predicate.call(this, this.input.peek())) {
      str = str + this.input.next()
    }

    return str
  }

  readNumber(): { type: 'num'; value: number } {
    let hasDot = false
    let n = this.readWhile(ch => {
      if (ch === '.') {
        if (hasDot) return false
        hasDot = true
        return true
      }
      return this.isDigit(ch)
    })

    return { type: 'num', value: Number.parseFloat(n) }
  }

  readIdentifier(): { type: 'kw' | 'var'; value: string } {
    let id = this.readWhile(this.isIdentifier)

    return { type: this.isKeyword(id) ? 'kw' : 'var', value: id }
  }

  readEscaped(end: string): string {
    let escaped = false
    let str = ''

    this.input.next()

    while (!this.input.eof()) {
      let ch = this.input.next()

      if (escaped) {
        str = str + ch
        escaped = false
      } else if (ch === '\\') {
        escaped = true
      } else if (ch === end) {
        break
      } else {
        str = str + ch
      }
    }

    return str
  }

  readString(): { type: 'str'; value: string } {
    return { type: 'str', value: this.readEscaped('"') }
  }

  skipComment(): void {
    this.readWhile(ch => ch !== '\n')
    this.input.next()
  }

  readNext(): Token | null {
    this.readWhile(this.isWhitespace)

    if (this.input.eof()) return null

    let ch = this.input.peek()

    if (ch === '#') {
      this.skipComment()
      return this.readNext()
    }

    if (ch === '"') return this.readString()
    if (this.isDigit(ch)) return this.readNumber()
    if (this.isIdentifierStart(ch)) return this.readIdentifier()
    if (this.isPunctuation(ch))
      return { type: 'punc', value: this.input.next() }
    if (this.isOperatorCharacter(ch))
      return { type: 'op', value: this.readWhile(this.isOperatorCharacter) }

    return this.input.croak(`Unable to handle character: ${ch}`)
  }

  peek() {
    return this.current || (this.current = this.readNext())
  }

  next() {
    let token = this.current
    this.current = null
    return token || this.readNext()
  }

  eof(): boolean {
    return this.peek() === null
  }

  croak(msg: string): never {
    return this.input.croak(msg)
  }
}
