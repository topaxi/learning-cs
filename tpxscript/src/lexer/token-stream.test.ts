import { InputStream } from '../input-stream'
import { TokenStream } from './token-stream'

function parse(str: string) {
  let i = new InputStream(str)
  let t = new TokenStream(i)

  let ast = []

  while (!t.eof()) {
    ast.push(t.next())
  }

  return ast
}

describe('tpxscript::TokenStream', () => {
  test('should tokenize numbers', () => {
    expect(parse('123.5')).toEqual([{ type: 'num', value: 123.5 }])
  })

  test('should tokenize strings', () => {
    expect(parse('"Hello, World!"')).toEqual([
      { type: 'str', value: 'Hello, World!' }
    ])
  })

  test('should tokenize booleans', () => {
    expect(parse('true')).toEqual([{ type: 'kw', value: 'true' }])
    expect(parse('false')).toEqual([{ type: 'kw', value: 'false' }])
  })

  test('should tokenize identifiers', () => {
    expect(parse('foo')).toEqual([{ type: 'var', value: 'foo' }])
  })

  test('should tokenize function declarations', () => {
    expect(parse('fn (x) x')).toMatchSnapshot()
  })

  test('should tokenize function calls', () => {
    expect(parse('foo(x, 1)')).toMatchSnapshot()
  })

  test('should tokenize conditionals', () => {
    expect(parse('if foo then bar else baz')).toMatchSnapshot()
    expect(parse('if foo then bar')).toMatchSnapshot()
  })

  test('should tokenize assignments', () => {
    expect(parse('x = 10')).toMatchSnapshot()
  })

  test('should tokenize arithmetic', () => {
    expect(parse('x + y')).toMatchSnapshot()
    expect(parse('x + y * z')).toMatchSnapshot()
  })

  test('should tokenize logical operators', () => {
    expect(parse('x && y')).toMatchSnapshot()
    expect(parse('x || y')).toMatchSnapshot()
    expect(parse('x and y')).toMatchSnapshot()
    expect(parse('x or y')).toMatchSnapshot()
  })

  test('should tokenize blocks', () => {
    expect(
      parse(`
      {
        a = 5;
        b = a * 2;
        a + b;
      }
    `)
    ).toMatchSnapshot()
  })

  test('should tokenize block variables', () => {
    expect(
      parse(`
      let (a = 10, b = a * 10) {
        a + b
      }
    `)
    ).toMatchSnapshot()
  })
})
