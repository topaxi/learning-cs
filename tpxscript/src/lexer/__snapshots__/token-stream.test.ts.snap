// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tpxscript::TokenStream should tokenize arithmetic 1`] = `
Array [
  Object {
    "type": "var",
    "value": "x",
  },
  Object {
    "type": "op",
    "value": "+",
  },
  Object {
    "type": "var",
    "value": "y",
  },
]
`;

exports[`tpxscript::TokenStream should tokenize arithmetic 2`] = `
Array [
  Object {
    "type": "var",
    "value": "x",
  },
  Object {
    "type": "op",
    "value": "+",
  },
  Object {
    "type": "var",
    "value": "y",
  },
  Object {
    "type": "op",
    "value": "*",
  },
  Object {
    "type": "var",
    "value": "z",
  },
]
`;

exports[`tpxscript::TokenStream should tokenize assignments 1`] = `
Array [
  Object {
    "type": "var",
    "value": "x",
  },
  Object {
    "type": "op",
    "value": "=",
  },
  Object {
    "type": "num",
    "value": 10,
  },
]
`;

exports[`tpxscript::TokenStream should tokenize block variables 1`] = `
Array [
  Object {
    "type": "var",
    "value": "let",
  },
  Object {
    "type": "punc",
    "value": "(",
  },
  Object {
    "type": "var",
    "value": "a",
  },
  Object {
    "type": "op",
    "value": "=",
  },
  Object {
    "type": "num",
    "value": 10,
  },
  Object {
    "type": "punc",
    "value": ",",
  },
  Object {
    "type": "var",
    "value": "b",
  },
  Object {
    "type": "op",
    "value": "=",
  },
  Object {
    "type": "var",
    "value": "a",
  },
  Object {
    "type": "op",
    "value": "*",
  },
  Object {
    "type": "num",
    "value": 10,
  },
  Object {
    "type": "punc",
    "value": ")",
  },
  Object {
    "type": "punc",
    "value": "{",
  },
  Object {
    "type": "var",
    "value": "a",
  },
  Object {
    "type": "op",
    "value": "+",
  },
  Object {
    "type": "var",
    "value": "b",
  },
  Object {
    "type": "punc",
    "value": "}",
  },
]
`;

exports[`tpxscript::TokenStream should tokenize blocks 1`] = `
Array [
  Object {
    "type": "punc",
    "value": "{",
  },
  Object {
    "type": "var",
    "value": "a",
  },
  Object {
    "type": "op",
    "value": "=",
  },
  Object {
    "type": "num",
    "value": 5,
  },
  Object {
    "type": "punc",
    "value": ";",
  },
  Object {
    "type": "var",
    "value": "b",
  },
  Object {
    "type": "op",
    "value": "=",
  },
  Object {
    "type": "var",
    "value": "a",
  },
  Object {
    "type": "op",
    "value": "*",
  },
  Object {
    "type": "num",
    "value": 2,
  },
  Object {
    "type": "punc",
    "value": ";",
  },
  Object {
    "type": "var",
    "value": "a",
  },
  Object {
    "type": "op",
    "value": "+",
  },
  Object {
    "type": "var",
    "value": "b",
  },
  Object {
    "type": "punc",
    "value": ";",
  },
  Object {
    "type": "punc",
    "value": "}",
  },
]
`;

exports[`tpxscript::TokenStream should tokenize conditionals 1`] = `
Array [
  Object {
    "type": "kw",
    "value": "if",
  },
  Object {
    "type": "var",
    "value": "foo",
  },
  Object {
    "type": "kw",
    "value": "then",
  },
  Object {
    "type": "var",
    "value": "bar",
  },
  Object {
    "type": "kw",
    "value": "else",
  },
  Object {
    "type": "var",
    "value": "baz",
  },
]
`;

exports[`tpxscript::TokenStream should tokenize conditionals 2`] = `
Array [
  Object {
    "type": "kw",
    "value": "if",
  },
  Object {
    "type": "var",
    "value": "foo",
  },
  Object {
    "type": "kw",
    "value": "then",
  },
  Object {
    "type": "var",
    "value": "bar",
  },
]
`;

exports[`tpxscript::TokenStream should tokenize function calls 1`] = `
Array [
  Object {
    "type": "var",
    "value": "foo",
  },
  Object {
    "type": "punc",
    "value": "(",
  },
  Object {
    "type": "var",
    "value": "x",
  },
  Object {
    "type": "punc",
    "value": ",",
  },
  Object {
    "type": "num",
    "value": 1,
  },
  Object {
    "type": "punc",
    "value": ")",
  },
]
`;

exports[`tpxscript::TokenStream should tokenize function declarations 1`] = `
Array [
  Object {
    "type": "kw",
    "value": "fn",
  },
  Object {
    "type": "punc",
    "value": "(",
  },
  Object {
    "type": "var",
    "value": "x",
  },
  Object {
    "type": "punc",
    "value": ")",
  },
  Object {
    "type": "var",
    "value": "x",
  },
]
`;

exports[`tpxscript::TokenStream should tokenize logical operators 1`] = `
Array [
  Object {
    "type": "var",
    "value": "x",
  },
  Object {
    "type": "op",
    "value": "&&",
  },
  Object {
    "type": "var",
    "value": "y",
  },
]
`;

exports[`tpxscript::TokenStream should tokenize logical operators 2`] = `
Array [
  Object {
    "type": "var",
    "value": "x",
  },
  Object {
    "type": "op",
    "value": "||",
  },
  Object {
    "type": "var",
    "value": "y",
  },
]
`;

exports[`tpxscript::TokenStream should tokenize logical operators 3`] = `
Array [
  Object {
    "type": "var",
    "value": "x",
  },
  Object {
    "type": "kw",
    "value": "and",
  },
  Object {
    "type": "var",
    "value": "y",
  },
]
`;

exports[`tpxscript::TokenStream should tokenize logical operators 4`] = `
Array [
  Object {
    "type": "var",
    "value": "x",
  },
  Object {
    "type": "kw",
    "value": "or",
  },
  Object {
    "type": "var",
    "value": "y",
  },
]
`;
